### PR TITLE
Minor coverage report changes

### PR DESCRIPTION
**Pull request by UlfNorell**
- Add coverage metadata for ignored locations that should not require code coverage.
- Bundle coverage data and index into `CoverageReport`
- Add some missing instances for coverage types

Required for input-output-hk/plutus-apps#416

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-tx/src/PlutusTx/Coverage.hs [hunk 1]
```diff
@@ -86,21 +91,24 @@ data CoverageAnnotation = CoverLocation CovLoc
...
                         deriving (Ord, Eq, Show, Read, Generic, Serialise)
                         deriving Flat via (AsSerialize CoverageAnnotation)
+                        deriving anyclass (NFData, ToJSON, FromJSON, ToJSONKey, FromJSONKey)
 
 instance Pretty CoverageAnnotation where
   pretty (CoverLocation loc) = pretty loc
   pretty (CoverBool loc b)   = pretty loc <+> "=" <+> pretty b
 
 data Metadata = ApplicationHeadSymbol String
+              | IgnoredAnnotation   -- ^ Location that is not interesting to cover.
```
Comment (michaelpj) [inline plutus-tx/src/PlutusTx/Coverage.hs:101]: I feel like I'm missing something - shouldn't you be introducing these somewhere?

Comment (UlfNorell) [reply plutus-tx/src/PlutusTx/Coverage.hs:101]: Yes, in plutus-apps.

**Comment by UlfNorell** [comment]
@michaelpj do you know what's up with the Hydra failure? I don't see how those can be my fault...

File: plutus-tx/src/PlutusTx/Coverage.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-tx/src/PlutusTx/Coverage.hs:100]: I'm still confused. These annotations are generated by the plugin, how are you adding them later?

Maybe you could include some documentation explaining how they're supposed to be used?

**Comment by michaelpj** [comment]
> do you know what's up with the Hydra failure?

no, it's weird and I don't like it.

File: plutus-tx/src/PlutusTx/Coverage.hs [hunk 2]
```diff
@@ -140,23 +155,52 @@ addBoolCaseToCoverageIndex src b meta = do
   tell $ CoverageIndex (Map.singleton ann meta)
   pure ann
 
+-- | Add metadata to a coverage annotation. Does nothing if the annotation is not in the index.
```
Comment (MaximilianAlgehed) [inline plutus-tx/src/PlutusTx/Coverage.hs:158]: @michaelpj c.f. on your question above.

File: plutus-tx/src/PlutusTx/Coverage.hs (hunk 1 above)
Comment (MaximilianAlgehed) [reply plutus-tx/src/PlutusTx/Coverage.hs]: Some context here: we are about to merge a static analysis tool into plutus-apps that adds these annotations post-fact.

**Comment by MaximilianAlgehed** [comment]
@michaelpj any progress on this? We are kindof blocked :(

**Comment by michaelpj** [comment]
No, I couldn't find the issue and I don't have time to look at it at the moment. Reluctantly, I suggest you comment out the tests that are failing in CI for now so we can move this on.

**Comment by MaximilianAlgehed** [comment]
@michaelpj a heads up that I've now ignored the tests that are causing CI to fail locally. Let's see what happens in hydra.

**Comment by MaximilianAlgehed** [comment]
Ok I officially have no clue what so ever.

**Comment by michaelpj** [comment]
Our Darwin Hydra builders are basically broken for weeks now. On master I've marked darwin as not required but you're based on an old commit apparently.

**Comment by michaelpj** [comment]
I'll just force merge, I can see that the linux jobs all completed.

**Comment by michaelpj** [comment]
Hang on, did this PR never go into master? That's problematic.

**Comment by michaelpj** [comment]
This is why we should only ever allow backports to release branches argh.

**Comment by michaelpj** [comment]
I will do a forward port.

**Pull request merged**